runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java

```diff
@@ -161,6 +179,40 @@ private static boolean isPrimitiveTransform(PTransform transform) {
     return network;
   }
 
+  // This enables a naive implementation of topological sort, instead of doing something clever.
+  // Nodes with lower weight precede nodes with higher weight, and are unrelated to nodes with
+  // equal weight
+  private final LoadingCache<PTransformNode, Long> nodeWeights =
+      CacheBuilder.newBuilder()
+          .build(
+              new CacheLoader<PTransformNode, Long>() {
+                @Override
+                public Long load(@Nonnull PTransformNode transformNode) {
+                  long parentWeight = 0L;
+                  for (String inputPCollectionId :
+                      transformNode.getTransform().getInputsMap().values()) {
+                    PTransformNode upstream =
+                        getProducer(
+                            PipelineNode.pCollection(
+                                inputPCollectionId,
+                                components.getPcollectionsOrThrow(inputPCollectionId)));
+                    parentWeight += nodeWeights.getUnchecked(upstream);
+                  }
+                  return 1 + parentWeight;
+                }
+              });
+
+  public Iterable<PTransformNode> getTopologicallyOrderedTransforms() {
+    return pipelineNetwork
+        .nodes()
+        .stream()
+        .filter(node -> node instanceof PTransformNode)
+        .map(PTransformNode.class::cast)
+        .collect(
+            Collectors.toCollection(
+                () -> new TreeSet<>(Comparator.comparingLong(nodeWeights::getUnchecked))));
```

Comment:
This comparator is really extremely inconsistent with equals, and will make it so there only looks like there's one root transform.